README.rst
Introduction
============

Provide a central URL to check the health of your Plone sites and for
warming up the cache.

The package has been written with HAProxy and performance in mind.

The provided view, ``@healthcheck`` will iterate over each Plone site
and load the frontpage and all linked resources.
It will also check the Plone site for direct descendent Navigation Roots.
If you have a multilingual site, you usually have a root folder for each language and this folder is a navigation root.
So this means, every language gets warmed up.

There is an alternative mode. If you call the view paths:list get parameters, healthcheck will check these pages, together with the resources this page references. There is no check if the path is INavigationRoot or an IPloneSite.
So this allows you to add any page to the check and the cache.

HAProxy calls the URL quite often. A cache mechanism ensures that by default:

  1. The check happens only between every 30-90 minutes. Inbetween you get a cached result
  2. If previous checks were successful, only one random site/navroot will be checked

The healthcheck will fail, if any navigation root page will fail to render.
There is only one exception. We will not fail on 401s. This may happen with pages that are private in nature.
They cannot be properly warmed up with this, but at least the healthcheck won't fail.
Resources referenced from the front page will NOT trigger an error.
Unfortunately, this is too often the case.

Comparison with collective.warmup
---------------------------------
This package was here first.

collective.warmup is a standalone script, existing outsite of zope/Plone.

Advantages of collective.warmup over pretaweb.healthcheck:

    - can be executed once during a deployment
    - Does (maybe?) not have the failure modes explained below

Advantages of pretaweb.healthcheck over collective.warmup:

    - Can be integrated in HAProxy
    - Caching included in pretaweb.healthcheck
    - Logs quite a few helpful information you can only get from within Zope, for example, how much of the cache is used after warmup

If you use HAProxy, you want to use pretaweb.healthcheck.

Installation
------------

Add the egg as a dependency to your policy egg.
This package does not store persistent configuration. There is no configuration option within the Plone site.

No installation or uninstallation profile needs to be executed.

Configuration
-------------

You can control cache times with environment variables:

    - `healthcheck_cache_interval`

      The default is 3600 seconds. You can define any value in seconds.

    - `healthcheck_cache_variance`

      the default is 1800 seconds. You can define any value in seconds.

To calculate the next expire time, the cache_interval is taken, and a random number between 0 and the cache_variance is added.
This time is calculated after every expiration.

If you do not want to warm up every single Plone site / navigation root, you can give paths to check for. These paths are checked **together** with the resources the page references::

    http://yourserver:port/@@healthcheck?paths:list=/Plone/en&paths:list=/Plone/en/a/b/c/d/expensive_page

Would load all resources of the english front page and the one expensive page deep down your site.

Failure modes
-------------

If you are going to integrate this package in your deployment together with HAProxy, be aware of the new failure modes.

If you have many Plones on a single Zope instance and a few of them are not actively used, they are still checked. If they produce failures after a minor Plone update, all Plone pages will not be available, because this package told HAProxy, that it itself is broken.

If you allow people to modify Plone pages through the web in such a way that the frontpage can become broken, your complete site can go down, but not necessary soonish.
the default cache expiration is 1 hour plus a random amount from half an hour. But if the previous check was successful, we don't need the cache warmup functionality, so we pick only a single page at random. If you think this is not a problem, do the math for yourself::

    >>> math.log(abs(probability - 1)) / math.log((sites - broken_sites) / float(sites)) * healthcheck_cache_interval * (healthcheck_cache_variance / 2)

If you have a zope instance with 20 Plone sites, after 17 hours you have a 50% chance that your instance gets identified as broken. If each instance has three language folders, the time increase to a bit more than two days. This is not a good start for debugging.

If you are in such a situation, you have a few possibilities:

  - Move your unused Plone pages into a subfolder. They won't be found any more. Earlier versions used the virtual_hosting mapping to find Plone sites. This version doesn't
  - Have proper monitoring in Place. `Sentry`_ has proper `Plone integration`_ and in its default configuration will send you mails for errors only and only once. So after cutting the noise by fixing a bunch of invisible bugs, every mail is usually an actionable item.
  - Configure this package and HAProxy so that you can use different backend for different Plone sites or language folders. Then the breakage is more local.

If you have a dedicated Zope Instance for a single Plone site, and your users can't break the Plone page easily, this is much less an issue.


Todo
----
- [ ] Caching under Plone 5 is currently not working very well: `subrequest bug`_ 

Testing
-------
The code has full test coverage for everything except the views.
To exercise the tests in a development checkout, run buildout, then run::

    $ ./bin/py.test --cov=pretaweb --cov-report=term-missing

.. _`subrequest bug`: https://github.com/Plone/Plone.subrequest/issues/6
.. _Plone integration: https://docs.getsentry.com/hosted/clients/python/integrations/zope/
.. _Sentry: https://www.getsentry.com


docs/HISTORY.rst
Changelog
=========


2.0.0 (unreleased)
------------------

- Rewritten to make testing simpler.
- Logging changed, some things got dropped, others got added
- Warm up navigation roots (Multilingual sites)

1.0 (2011-06-24)
----------------

- Removed force=yes option to prevent DoS attacks 
- Output now to instance stderr


0.1dev (unreleased)
-------------------

- Plone instance discovery / ZODB traveersal
- Resource sub-requesting
- Page xml interigation and sub-requesting
- CSS interigation and sub-requesting
- Second level page sub-requesting
- Added verbose=yes for better output control
- Added force=yes to force a re-check
- Added ignoreResourceServerError=yes option to ignore 5xx errors returned by
  sub-resources




tests/test_health_check.py
from datetime import datetime
from datetime import timedelta as td
from pretaweb.healthcheck.check import HealthCheck
from pretaweb.healthcheck.check import NotExpired
from pretaweb.healthcheck.check import STATUS_ERROR
from pretaweb.healthcheck.check import STATUS_HEALTHY
from ZODB.POSException import ConflictError

import pytest


now = datetime.utcnow()

quite_old = now - td(days=100)

ten_minutes_from_now = now + td(seconds=600)


def is_between_20_to_100_minutes_from_now(testtime):
    return (now + td(seconds=20*60)) < testtime < (now + td(seconds=100*60))


def test_first_run():
    checker = HealthCheck(last_result=None,
                          expire_time=quite_old,
                          traverser=None,
                          context=None,
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )
    new_expire, result = checker()
    assert is_between_20_to_100_minutes_from_now(new_expire)


def test_cached_run():
    last_result = 'same'
    checker = HealthCheck(last_result=last_result,
                          expire_time=ten_minutes_from_now,
                          traverser=None,
                          context=None,
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )
    pytest.raises(NotExpired, checker)


def test_avoid_parallel_runs(monkeypatch):
    monkeypatch.setattr('pretaweb.healthcheck.check.AM_I_RUNNING', True)
    last_result = 'same'
    checker = HealthCheck(last_result=last_result,
                          expire_time=ten_minutes_from_now,
                          traverser=None,
                          context=None,
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )
    pytest.raises(NotExpired, checker)


def test_recheck_checks_one():
    checker = HealthCheck(last_result=STATUS_HEALTHY,
                          expire_time=quite_old,
                          traverser=None,
                          context=None,
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )

    def _get_pages():
        return (1, 2)

    remember_them = []

    def _wake_plone(a_plone):
        remember_them.append(a_plone)

    checker._get_pages = _get_pages
    checker._wake_plone = _wake_plone

    new_expire, result = checker()
    assert is_between_20_to_100_minutes_from_now(new_expire)
    assert result == STATUS_HEALTHY

    assert len(remember_them) == 1
    assert remember_them[0] in (1, 2)


def test_full_check_after_unhealthy_check():
    checker = HealthCheck(last_result=STATUS_ERROR,
                          expire_time=quite_old,
                          traverser=None,
                          context=None,
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )

    def _get_pages():
        return (1, 2)

    remember_them = []

    def _wake_plone(a_plone):
        remember_them.append(a_plone)

    checker._get_pages = _get_pages
    checker._wake_plone = _wake_plone

    new_expire, result = checker()
    assert is_between_20_to_100_minutes_from_now(new_expire)
    assert result == STATUS_HEALTHY

    assert len(remember_them) == 2


def test_check_does_not_handle_conflict_error():
    checker = HealthCheck(last_result=STATUS_ERROR,
                          expire_time=quite_old,
                          traverser=None,
                          context=None,
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )

    def _get_pages():
        raise ConflictError()

    checker._get_pages = _get_pages

    pytest.raises(ConflictError, checker)


def test_check_does_handle_other_exceptions():
    checker = HealthCheck(last_result=STATUS_ERROR,
                          expire_time=quite_old,
                          traverser=None,
                          context=None,
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )

    def _get_pages():
        raise KeyError()

    checker._get_pages = _get_pages

    new_expire, result = checker()
    assert is_between_20_to_100_minutes_from_now(new_expire)
    assert result == STATUS_ERROR


def test_get_pages_handles_empty():
    checker = HealthCheck(last_result=STATUS_ERROR,
                          expire_time=quite_old,
                          traverser=None,
                          context={},
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )

    assert [] == list(checker._get_pages())


class FakeSiteRoot(object):
    def __init__(self, replies):
        self.replies = replies

    def providedBy(self, whatever):
        return self.replies.pop()


def test_get_plone_navroots(monkeypatch):
    monkeypatch.setattr('pretaweb.healthcheck.check.INavigationRoot',
                        FakeSiteRoot([True, True]))
    checker = HealthCheck(last_result=STATUS_ERROR,
                          expire_time=quite_old,
                          traverser=None,
                          context={'a': 'a', 'b': 'b'},
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )

    plones = list(checker._get_pages())
    assert ['a', 'b'] == plones


def test_get_plone_site_root_and_nav_root(monkeypatch):
    monkeypatch.setattr('pretaweb.healthcheck.check.IPloneSiteRoot',
                        FakeSiteRoot([False, True]))
    monkeypatch.setattr('pretaweb.healthcheck.check.INavigationRoot',
                        FakeSiteRoot([False, True]))
    checker = HealthCheck(last_result=STATUS_ERROR,
                          expire_time=quite_old,
                          traverser=None,
                          context={'a': {'myplone': True}},
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )

    plones = list(checker._get_pages())
    assert [{'myplone': True}] == plones


def test_get_plone_navroots_multilingual(monkeypatch):
    monkeypatch.setattr('pretaweb.healthcheck.check.IPloneSiteRoot',
                        FakeSiteRoot([False, True]))
    monkeypatch.setattr('pretaweb.healthcheck.check.INavigationRoot',
                        FakeSiteRoot([True, True]))
    checker = HealthCheck(last_result=STATUS_ERROR,
                          expire_time=quite_old,
                          traverser=None,
                          context={'a': {'de': 'de_folder'}},
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )

    plones = list(checker._get_pages())
    assert [{'de': 'de_folder'}, 'de_folder'] == plones


class FakeContext(object):
    def unrestrictedTraverse(self, path):
        if 'missing' in path:
            raise KeyError
        return path


def test_get_plone_from_paths():
    checker = HealthCheck(last_result=STATUS_ERROR,
                          expire_time=quite_old,
                          traverser=None,
                          context=FakeContext(),
                          base=None,
                          host=None,
                          use_https=False,
                          paths=['/a', '/missing'],
                          )

    plones = list(checker._get_pages())
    assert ['/a'] == plones


class FakePloneObj(object):
    def getPhysicalPath(self):
        return ('a', 'b', 'c')


def test_wake_plone(monkeypatch):
    urls_gotten = []

    class PloneLoader(object):
        def __init__(self, base, host, use_https, url):
            urls_gotten.append(url)

        def __call__(self):
            pass

    monkeypatch.setattr('pretaweb.healthcheck.check.PloneLoader',
                        PloneLoader)

    checker = HealthCheck(last_result=STATUS_ERROR,
                          expire_time=quite_old,
                          traverser=lambda x: x,
                          context=None,
                          base=None,
                          host=None,
                          use_https=False,
                          paths=None,
                          )
    checker._wake_plone(FakePloneObj())

    assert ['a/b/c'] == urls_gotten


tests/test_plone_loader.py
from pretaweb.healthcheck.check import PloneLoader
import pytest


def make_subrequester(monkeypatch, status, response, **special_replies):
    memory = []

    class Response(object):
        def __init__(self, url, status, headers, body):
            self.url = url
            self.status = status
            self.body = body
            self.headers = headers

        def getStatus(self):
            return self.status

        def getHeader(self, key):
            try:
                return self.headers[key.lower()]
            except KeyError:
                raise KeyError('Key "%r" not found for resource "%r"',
                               key.lower(), self.url)

        def getBody(self):
            return self.body

    def subrequest(url):
        memory.append(url)
        if url in special_replies:
            return Response(url, *special_replies[url])
        else:
            return Response(url, status, {}, response)
    monkeypatch.setattr('pretaweb.healthcheck.check.subrequest', subrequest)
    return memory


def test_plone_loader(monkeypatch):
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')
    subrequest_memory = make_subrequester(monkeypatch, 200, '')
    loader()
    assert ['http://example.com'] == subrequest_memory


def test_plone_loader_no_fail_on_401(monkeypatch):
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')
    make_subrequester(monkeypatch, 401, '')
    loader()
    assert True


def test_plone_loader_failed(monkeypatch):
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')
    make_subrequester(monkeypatch, 599, '')
    try:
        loader()
        assert False
    except Exception, e:
        assert 'http://example.com' == e.url
        assert 599 == e.status


def test_plone_loader_with_urls(monkeypatch):
    plone_body = '''
<html>
  <body>
    <div>
      <a href="/a"></a>
      <a href="#xx"></a>
      <img src="/b"></img>
      <script src="/c"></script>
    </div>
  </body>
</html>
'''
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')
    subrequest_memory = make_subrequester(monkeypatch, 200, plone_body)
    loader()
    assert ['http://example.com',
            '/a',
            '/b',
            '/c'] == subrequest_memory


@pytest.mark.parametrize('url,status_code,expected', (
    ('/a', 200, ['/a']),
    ('/a', 299, ['/a']),
    ('/a', 300, ['/a']),
    ('/a', 399, ['/a']),
    ('/a', 400, ['/a']),
    ('/a', 499, ['/a']),
    ('/a', 500, ['/a']),
    ('/a', 599, ['/a']),
))
def test_wake_resource(monkeypatch, status_code, url, expected):
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')
    subrequest_memory = make_subrequester(monkeypatch, 200, 'xx',
                                          **{url: (status_code, {}, '')})

    loader._wake_resource(url)
    loader._wake_resource(url)

    assert expected == subrequest_memory


@pytest.mark.parametrize('url,content,content_type,expected', (
    ('/a', 'url(/image)', 'text/css', ['/a', '/image']),
    ('/a.css', 'url(/image)', 'text/html', ['/a.css', '/image']),
    ('/a.css', 'url(/image)', 'text/css', ['/a.css', '/image']),
    ('/a.html', 'url(/image)', 'text/html', ['/a.html']),
    ('/a.kss', 'url(/image)', 'text/css', ['/a.kss']),
    ('/a', '', 'text/css', ['/a']),
    ('/a.css', 'url(#xx)', 'text/css', ['/a.css']),
))
def test_wake_css_resource(monkeypatch, url, content, content_type,
                           expected):
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')
    other_urls = {url: (200, {'content-type': content_type}, content),
                  '/image': (200, {}, '')}
    subrequest_memory = make_subrequester(monkeypatch, 200, content,
                                          **other_urls)

    loader._wake_css_resource(url)
    loader._wake_css_resource(url)

    assert expected == subrequest_memory


def test_plone_loader_finds_css(monkeypatch):
    plone_body = '''
<html>
  <head>
    <link href="/a" />
    <link href="#xx" />
    <link href="/b" />
    <link href="/c.css" />
    <link href="/d" />
    <link href="/e" />
    <link href="/f.kss" />
    <style>
      @import url("/x");
      @import url("#xx");
  <body>
  </body>
</html>
'''
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')
    ignore_link = 'url(/dontfollow)'
    follow_link = 'url(/follow_me)'
    bad_link = 'url(#xx)'
    resources = {'/a': (200, {'content-type': 'text/css'}, follow_link),
                 '/b': (200, {'content-type': 'text/css'}, ''),
                 '/c.css': (200, {'content-type': 'text/html'}, follow_link),
                 '/d': (200, {'content-type': 'text/html'}, ignore_link),
                 '/e': (200, {'content-type': 'text/css'}, bad_link),
                 '/f.kss': (200, {'content-type': 'text/css'}, ignore_link),
                 '/x': (200, {'content-type': 'text/css'}, follow_link),
                 }
    # a: Link should be followed, because ct is right
    # b: should be followed but has no content
    # c: should be followed because ends with .css
    # d: should not be followed, because no indication of css
    # e: should not be followed, because indication that it is kss
    # x: should be followed for same reasons as a
    subrequest_memory = make_subrequester(monkeypatch, 200, plone_body,
                                          **resources)
    loader()
    assert ['http://example.com',
            '/a',
            '/follow_me',
            '/b',
            '/c.css',
            '/d',
            '/e',
            '/f.kss',
            '/x',
            ] == subrequest_memory


@pytest.mark.parametrize('url,expected', (
    ('#x', None),
    ('', None),
    ('/ex', None),
    ('http://example.com/x', '/x'),
    ('https://example.com/x', None),
    ('bla', '/ex/bla'),
))
def test_normalize_link(url, expected):
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')

    assert expected == loader._normalize_link('http://example.com/ex', url)


def test_normalize_link_https():
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')

    assert '/ex2' == loader._normalize_link('https://example.com/ex', '/ex2')


def test_normalize_link_relative_ref_link():
    loader = PloneLoader('http://example.com',
                         'example.com',
                         False,
                         'http://example.com')

    assert '/ex/ex2' == loader._normalize_link('/ex', 'ex2')


pretaweb/healthcheck/views.py
from datetime import datetime
from datetime import timedelta
from pretaweb.healthcheck.check import HealthCheck
from pretaweb.healthcheck.check import STATUS_ERROR
from pretaweb.healthcheck.check import NotExpired
from Products.Five import BrowserView

import logging


logger = logging.getLogger(__name__)

CACHED_HEALTH_CHECK_RESULT = STATUS_ERROR
HEALTH_CHECK_NEXT_EXPIRE = datetime.utcnow() - timedelta(seconds=1)

CHECK_IN_PROGRESS = False


class HealthCheckView(BrowserView):

    def __call__(self, paths=None):
        environ = self.request.environ
        context = self.context

        global CACHED_HEALTH_CHECK_RESULT
        global HEALTH_CHECK_NEXT_EXPIRE
        global CHECK_IN_PROGRESS

        if CHECK_IN_PROGRESS and datetime.utcnow() < HEALTH_CHECK_NEXT_EXPIRE:
            logger.info('Check in progress, ignoring returning old result')
            return self._build_response(CACHED_HEALTH_CHECK_RESULT)
        else:
            CHECK_IN_PROGRESS = True

        use_https = environ.get('HTTPS', False)

        db = self.context._p_jar.db()

        cache_percent_mod = float(db.getCacheSize()) / 100
        cache_utiliziation_before = db.cacheSize() / cache_percent_mod

        health_check = HealthCheck(last_result=CACHED_HEALTH_CHECK_RESULT,
                                   expire_time=HEALTH_CHECK_NEXT_EXPIRE,
                                   traverser=context.restrictedTraverse,
                                   context=self.context,
                                   base=self.request.base,
                                   host=environ.get('HTTP_HOST', ''),
                                   use_https=use_https,
                                   paths=paths,
                                   )

        try:
            HEALTH_CHECK_NEXT_EXPIRE, CACHED_HEALTH_CHECK_RESULT = \
                health_check()
        except NotExpired:
            return self._build_response(CACHED_HEALTH_CHECK_RESULT)

        cache_utiliziation_after = db.cacheSize() / cache_percent_mod

        # Get status
        status = CACHED_HEALTH_CHECK_RESULT
        CHECK_IN_PROGRESS = False

        logger.info('healthcheck result: %r', status)
        logger.info('Cache fill level before: %.2f %%. After: %.2f %%.',
                    cache_utiliziation_before, cache_utiliziation_after)

        return self._build_response(status)

    def _build_response(self, status):
        # Construst Response
        response = self.request.response
        response.setStatus(status)
        response.setHeader('Content-type:', 'text/plain')
        responseLine = '%s %s\n' % (
            status,
            {200: 'OK', 503: 'Service Unavailable'}.get(status, ''))

        return responseLine


pretaweb/healthcheck/check.py
from datetime import datetime
from datetime import timedelta
from lxml import etree
from os import getenv
from plone.app.layout.navigation.interfaces import INavigationRoot
from plone.subrequest import subrequest
from Products.CMFPlone.interfaces.siteroot import IPloneSiteRoot
from random import choice
from random import randint
from StringIO import StringIO
from time import time
from ZODB.POSException import ConflictError

import logging
import re
import urllib


logger = logging.getLogger(__name__)

STATUS_ERROR = 500
STATUS_HEALTHY = 200

CSS_IMPORT_RE = re.compile(r'@import\s+url\([\'"]{0,1}([^)]+?)[\'"]{0,1}\)')

INTERVAL = int(getenv('healthcheck_cache_interval', 3600))
INTERVAL_VARIANCE = int(getenv('healthcheck_cache_variance',  1800))


AM_I_RUNNING = False


class RequestError(Exception):
    def __init__(self, url, status):
        self.url = url
        self.status = status


class ServerError(RequestError):
    pass


class NotExpired(Exception):
    pass


class PloneLoader(object):
    def __init__(self, base, host, uses_https, url):
        self.start_url = url
        self.base = base
        self.host = host
        self.uses_https = uses_https
        self.urls = set()

    def _request(self, url, raises):
        logger.info('Calling %r',
                    url)
        response = subrequest(url)
        status = response.getStatus()
        if status >= 400:
            raise RequestError(url=url, status=status)
        return response

    def _wake_resource(self, url, start_url=None):
        if url in self.urls:
            return
        if not start_url:
            start_url = self.start_url
        self._lastRequestedURL = url
        response = subrequest(url)
        status = response.getStatus()
        if status == 200:
            logger.debug('Successfully loaded %r',
                         url)
            self.urls.add(url)
            return response

        elif 300 <= status < 400:
            logger.info('Page %r contains a reference on %r '
                        'but redirects with status code %r. '
                        'The Resource was not woken up. Fix your page',
                        start_url, url, status)
        else:
            logger.error('Page %r contains a reference to %r. '
                         'Trying to load the resource triggered an error '
                         'with status code %r',
                         start_url, url, status)
        self.urls.add(url)
        return None

    def _wake_css_resource(self, url):
        cssURLResourcePattern = re.compile(r"url\('?([^')]+)'?\)")

        # Get css resources and parse for url(...) directives

        response = self._wake_resource(url)
        if not response:
            return
        body = response.getBody()
        if not body:
            return

        if not (url.endswith('.css') or
                response.getHeader('content-type').startswith('text/css')):
            return
        if url.endswith('.kss'):
            return
        # get CSS working path
        workingPath = url.split('/')
        workingPath.pop()
        workingPath = self.base + '/'.join(workingPath)

        # parse URLs
        for match in cssURLResourcePattern.finditer(body):
            groups = match.groups()
            if len(groups) > 0:
                ref_url = '/'.join(url.split('/')[:-1])
                normalized_link = self._normalize_link(ref_url, groups[0])
                if normalized_link is None:
                    continue
                self._wake_resource(normalized_link, url)

    def _normalize_link(self, reference_url, link):
        base = self.base

        protocol = 'https' if self.uses_https else 'http'

        link = link.split('#')[0]
        if not len(link):
            return None
        link = urllib.unquote(link)
        if link[0] == '/':
            link = protocol + '://' + self.host + link
        link = link.rstrip('/')

        # Link same as base - ignore
        if reference_url == link:
            return None

        # Absolute URL - add without base
        if link.startswith(base):
            p = link[len(base):]
            return p

        # Relative URL - add with workingURL
        elif not (link.startswith('http://') or link.startswith('https://')):
            ll = reference_url + '/' + link
            if reference_url.startswith(base):
                return ll[len(base):]
            return ll

        else:
            logger.debug('\tResource out of scope: %s', link)

        return None

    def __call__(self):
        try:
            response = self._request(self.start_url, raises=True)
        except RequestError, e:
            if e.status == 401:
                logger.info('I tried to warm up a Plone site %r that '
                            'is private',
                            self.start_url)
                return
            else:
                raise
        body = response.getBody()

        logger.info('Plone site %s, Status %s',
                    self.start_url, response.getStatus())

        try:
            tree = etree.parse(StringIO(body), etree.HTMLParser())
        except etree.XMLSyntaxError:
            logger.warning('Error parsing the front page', exc_info=True)
            return

        self.process_body(body, tree)

    def process_body(self, body, tree):
        normal_links = tree.xpath('/html/body//a/@href|'
                                  '/html/body//img/@src|'
                                  '/html//script/@src')
        for link in normal_links:
            normalized_link = self._normalize_link(self.start_url, link)
            if normalized_link is None:
                continue
            self._wake_resource(normalized_link)

        for css_link in tree.xpath('/html//link/@href'):
            normalized_link = self._normalize_link(self.start_url, css_link)
            if normalized_link is None:
                continue
            self._wake_css_resource(normalized_link)

        for match in CSS_IMPORT_RE.finditer(body):
            groups = match.groups()
            if len(groups) > 0:
                normalized_link = self._normalize_link(self.start_url,
                                                       groups[0])
                if normalized_link is None:
                    continue
                self._wake_css_resource(normalized_link)


class HealthCheck(object):
    """
    Query URLs, recursivly.
    """
    def __init__(self,
                 last_result,
                 expire_time,
                 context,
                 traverser,
                 base,
                 host,
                 use_https,
                 paths=None,
                 ):
        self.last_result = last_result
        self.expire_time = expire_time
        self.context = context
        self.traverser = traverser
        self.base = base
        self.host = host
        self.use_https = use_https
        self.paths = paths

    def __call__(self):
        global AM_I_RUNNING
        try:
            if AM_I_RUNNING and datetime.utcnow() < self._next_expire():
                logger.info('I am still doing the health check, but just '
                            'got asked again. I will return the last state.')
                raise NotExpired()
            AM_I_RUNNING = True
            if datetime.utcnow() < self.expire_time:
                raise NotExpired()
            start = time()

            try:
                if self.last_result == STATUS_HEALTHY:
                    logger.info('Doing limited recheck')
                    self._wake_plone(choice(list(self._get_pages())))
                else:
                    logger.info('Doing full check')
                    for plone in self._get_pages():
                        self._wake_plone(plone)
            except Exception, e:
                if isinstance(e, ConflictError):
                    raise
                logger.exception('Healthcheck found a problem')
                # Instance no longer healthy
                result = STATUS_ERROR
            else:
                logger.info('Finished health check in %i seconds. Passed.',
                            time() - start)
                result = STATUS_HEALTHY

            new_expire = self._next_expire()

            logger.info('Next health check in %s',
                        new_expire.isoformat())

            return new_expire, result
        finally:
            AM_I_RUNNING = False

    def _next_expire(self):
        new_expire_in = INTERVAL + randint(0, INTERVAL_VARIANCE)
        return datetime.utcnow() + timedelta(seconds=new_expire_in)

    def _get_pages(self):
        logger.debug('Checking a list of given paths')
        if self.paths:
            return self._get_path_filtered_pages()
        else:
            return self._get_all_plone_and_navroots()

    def _get_path_filtered_pages(self):
        logger.debug('Checking all toplevel plone pages and their '
                     'direct nav root pages')
        for path in self.paths:
            try:
                yield self.context.unrestrictedTraverse(path)
            except KeyError:
                logger.warning(('I was asked to to a halth check for %r '
                                'but I cannot find an object on this '
                                'path. I ignore this'), path)
                continue

    def _get_all_plone_and_navroots(self):
        for item in self.context.values():
            if IPloneSiteRoot.providedBy(item):
                if INavigationRoot.providedBy(item):
                    yield item
                for sub_item in item.values():
                    if INavigationRoot.providedBy(sub_item):
                        yield sub_item
            elif INavigationRoot.providedBy(item):
                yield item

    def _wake_plone(self, plone):
        url = '/'.join(plone.getPhysicalPath())
        plone_loader = PloneLoader(self.base, self.host, self.use_https, url)
        plone_loader()
